src/main/java/org/truffleruby/extra/TruffleGraalNodes.java

```diff
@@ -109,12 +112,15 @@ public DynamicObject copyCapturedLocals(DynamicObject proc) {
                 callTargetForLambdas = Layouts.PROC.getCallTargetForLambdas(proc);
             }
 
+            final Object[] args = RubyArguments.pack(null, null, RubyArguments.getMethod(declarationFrame), null, nil(), null, new Object[0]);
+            final MaterializedFrame newDeclarationFrame = Truffle.getRuntime().createMaterializedFrame(args, descriptor);
+
             return coreLibrary().getProcFactory().newInstance(Layouts.PROC.build(
                     Layouts.PROC.getType(proc),
                     Layouts.PROC.getSharedMethodInfo(proc),
                     newCallTarget,
                     callTargetForLambdas,
-                    null, // The Proc no longer needs a declaration frame
+                    newDeclarationFrame, // The Proc no longer needs a original declaration frame
```

Comment:
Could you move the comments above `newDeclarationFrame =` and change it something like:
```
// The Proc no longer needs a declaration frame.
// However all Proc should have a declaration frame (for Proc#binding) so we create an empty one.
```